lab-evan-erick-michael/lib/bmp-reader.js

```diff
@@ -0,0 +1,196 @@
+      let red = 255 - pixelPalette[0];
+      let green = 255 - pixelPalette[1];
+      let blue = 255 - pixelPalette[2];
+
+      inverted.invertedPalette[i] = [red, green, blue, 0];
+    }
+    return inverted.invertedPalette;
+  }
+
+  function transformScale() {
+    for(var i=0;i<colorScaling.scaledPalette.length;i++) {
+      let red = bufferData.readUInt8(i);
+      red/=255;
+      red = Math.round(red*=100);
+      let green = bufferData.readUInt8(i + 1);
+      green/=255;
+      green = Math.round(green*=100);
+      let blue = bufferData.readUInt8(i + 2);
+      blue/=255;
+      blue = Math.round(blue*=100);
+      colorScaling.scaledPalette[i] = [red, green, blue, 0];
+    }
+    return colorScaling.scaledPalette;
+  }
+
+  getMaxVal();
+  readPalette();
+  transformGray();
+  randomTransform();
+  invertedTransform();
+  transformScale();
+
+  //call transform function, store array in variable, make one array with all nested arrays
+  gray.grayPalette.forEach(function(item) {
+    item.forEach(function(item2) {
+      gray.transformedGrayPalette.push(item2);
+    });
+  });
+
+  inverted.invertedPalette.forEach(function(item) {
+    item.forEach(function(item2) {
+      inverted.transformedPaletteInverted.push(item2);
+    });
+  });
+
+  random.randomPalette.forEach(function(item) {
+    item.forEach(function(item2) {
+      random.transformedRandomPalette.push(item2);
+    });
+  });
+
+  colorScaling.scaledPalette.forEach(function(item) {
+    item.forEach(function(item2) {
+      colorScaling.transformedScaledPalette.push(item2);
+    });
+  });
+
+  //create new buffer of transformed color palette
+  var grayBuffer = new Buffer(gray.transformedGrayPalette);
+  var invertedBuffer = new Buffer(inverted.transformedPaletteInverted);
+  var randomBuffer = new Buffer(random.transformedRandomPalette);
+  var colorScalingBuffer = new Buffer(colorScaling.transformedScaledPalette);
+  var header = bufferData.slice(0, 54);
+  var tail = bufferData.slice(1078);
+  //concat header, transform buffer and tail to read to file
+  var newBufferGray = Buffer.concat([header, grayBuffer, tail]);
+  var newBufferInverted = Buffer.concat([header, invertedBuffer, tail]);
+  var newBufferRandom = Buffer.concat([header, randomBuffer, tail]);
+  var newBufferScaled = Buffer.concat([header, colorScalingBuffer, tail]);
+  //write new buffer to file
+  fs.writeFile('../img/palette-bmp.bmp', newBufferRandom, (err) => {
+    if (err) {
+      console.log('error', err);
+    }
+    fs.writeFile('../img/palette-inverted.bmp', newBufferInverted, (err) => {
+      if (err) {
+        console.log('error', err);
+      }
+      fs.writeFile('../img/colorScale-bmp.bmp', newBufferScaled, (err) => {
+        if (err) {
+          console.log('error', err);
+        }
+        fs.writeFile('../img/grayscale-bmp.bmp', newBufferGray, (err) => {
+          if (err) {
+            console.log('error', err);
+          }
+        });
+      });
+    });
+  });
+});
```

Comment:
@kcirekcom @jmpaik @EWPS07 overall, the `bitmapReader` module functionality looks good - that said, there is a substantial amount of repeated code along with a series of antipatterns and related functions that could be refactored to be contained using the constructor pattern and a few prototype methods - let's set aside 15 minutes or so to chat after the holiday break and I'll give you some pointers on how to refactor this and clean it up